From c74923972c3efdd9e32015292db1aafc7836e695 Mon Sep 17 00:00:00 2001
From: Robert Siebert <rsiebert@miroku.no-ip.com>
Date: Wed, 20 Mar 2013 15:34:06 +0100
Subject: [PATCH 056/760] Added clear database option. This will remove all
 contents from the two gallery related tables. The
 user then needs to resynchronize. The button is
 optional in a theme.

Signed-off-by: Stuart Morgan <smorgan@mythtv.org>
---
 mythtv/programs/mythfrontend/galleryconfig.cpp     |   96 ++++++++++++++++----
 mythtv/programs/mythfrontend/galleryconfig.h       |    4 +
 .../mythfrontend/gallerydatabasehelper.cpp         |   20 ++++
 .../programs/mythfrontend/gallerydatabasehelper.h  |    2 +
 4 files changed, 106 insertions(+), 16 deletions(-)

diff --git a/mythtv/programs/mythfrontend/galleryconfig.cpp b/mythtv/programs/mythfrontend/galleryconfig.cpp
index 1dba64f..be7a438 100644
--- a/mythtv/programs/mythfrontend/galleryconfig.cpp
+++ b/mythtv/programs/mythfrontend/galleryconfig.cpp
@@ -8,27 +8,15 @@
 #include "mythuibutton.h"
 #include "mythuibuttonlist.h"
 #include "mythuispinbox.h"
+#include "mythdialogbox.h"
+#include "mythscreentype.h"
 
 #include "galleryconfig.h"
+#include "gallerydatabasehelper.h"
 #include "gallerytypedefs.h"
 
 
 
-enum FileSortOrder {
-    kSortByNameAsc     = 0,
-    kSortByNameDesc    = 1,
-    kSortByModTimeAsc  = 2,
-    kSortByModTimeDesc = 3,
-    kSortByExtAsc      = 4,
-    kSortByExtDesc     = 5,
-    kSortBySizeAsc     = 6,
-    kSortBySizeDesc    = 7,
-    kSortByDateAsc     = 8,
-    kSortByDateDesc    = 9
-};
-
-
-
 /** \fn     GalleryConfig::GalleryConfig(MythScreenStack *, const char *)
  *  \brief  Constructor
  *  \param  parent The screen parent
@@ -44,7 +32,8 @@ GalleryConfig::GalleryConfig(MythScreenStack *parent, const char *name)
       m_transitionTime(NULL),
       m_showHiddenFiles(NULL),
       m_saveButton(NULL),
-      m_cancelButton(NULL)
+      m_cancelButton(NULL),
+      m_clearDbButton(NULL)
 {
     // preset or load all variables
     m_sortOrder = 0;
@@ -83,6 +72,7 @@ bool GalleryConfig::Create()
 
     UIUtilE::Assign(this, m_saveButton, "save", &err);
     UIUtilE::Assign(this, m_cancelButton, "cancel", &err);
+    UIUtilW::Assign(this, m_clearDbButton, "cleardatabase", &err);
 
     // check if all widgets are present
     if (err)
@@ -98,6 +88,9 @@ bool GalleryConfig::Create()
     connect(m_saveButton, SIGNAL(Clicked()), this, SLOT(Save()));
     connect(m_cancelButton, SIGNAL(Clicked()), this, SLOT(Exit()));
 
+    if (m_clearDbButton)
+        connect(m_clearDbButton, SIGNAL(Clicked()), this, SLOT(ConfirmClearDatabase()));
+
     BuildFocusList();
 
     SetFocusWidget(m_storageGroupName);
@@ -128,6 +121,37 @@ bool GalleryConfig::keyPressEvent(QKeyEvent *event)
 
 
 
+/** \fn     GalleryView::customEvent(QEvent *)
+ *  \brief  Translates the keypresses to specific actions within the plugin
+ *  \param  event The custom event
+ *  \return void
+ */
+void GalleryConfig::customEvent(QEvent *event)
+{
+    if (event->type() == DialogCompletionEvent::kEventType)
+    {
+        DialogCompletionEvent *dce = (DialogCompletionEvent*)(event);
+
+        QString resultid  = dce->GetId();
+        int     buttonnum = dce->GetResult();
+
+        // Confirm current file deletion
+        if (resultid == "confirmdelete")
+        {
+            switch (buttonnum)
+            {
+            case 0 :
+                break;
+            case 1 :
+                ClearDatabase();
+                break;
+            }
+        }
+    }
+}
+
+
+
 /** \fn     GalleryConfig::Load()
  *  \brief  Load the values from the database and adds them to the widgets
  *  \return void
@@ -212,3 +236,43 @@ void GalleryConfig::Exit()
 {
     Close();
 }
+
+
+
+/** \fn     GalleryConfig::ConfirmClearDatabase()
+ *  \brief  Asks the user to confirm the removal of
+ *          all image related contents from the database
+ *  \return void
+ */
+void GalleryConfig::ConfirmClearDatabase()
+{
+    QString msg = QString("Do you want to clear all database contents?");
+    MythScreenStack         *m_popupStack = GetMythMainWindow()->GetStack("popup stack");
+    MythConfirmationDialog  *m_confirmPopup = new MythConfirmationDialog(m_popupStack, msg, true);
+
+    if (m_confirmPopup->Create())
+    {
+        m_confirmPopup->SetReturnEvent(this, "confirmdelete");
+        m_popupStack->AddScreen(m_confirmPopup);
+    }
+    else
+        delete m_confirmPopup;
+}
+
+
+
+/** \fn     GalleryConfig::ClearDatabase()
+ *  \brief  Clears all image related contents from the database
+ *  \return void
+ */
+void GalleryConfig::ClearDatabase()
+{
+    GalleryDatabaseHelper *m_dbHelper = new GalleryDatabaseHelper();
+    m_dbHelper->ClearDatabase();
+    delete m_dbHelper;
+
+    // tell the main view to reload the images
+    // because the storage group dir might have changed
+    emit configSaved();
+    Close();
+}
diff --git a/mythtv/programs/mythfrontend/galleryconfig.h b/mythtv/programs/mythfrontend/galleryconfig.h
index f955593..d3d471e 100644
--- a/mythtv/programs/mythfrontend/galleryconfig.h
+++ b/mythtv/programs/mythfrontend/galleryconfig.h
@@ -17,6 +17,7 @@ public:
 
     bool Create();
     bool keyPressEvent(QKeyEvent *);
+    void customEvent(QEvent*);
 
 signals:
     void configSaved();
@@ -31,11 +32,14 @@ private:
 
     MythUIButton       *m_saveButton;
     MythUIButton       *m_cancelButton;
+    MythUIButton       *m_clearDbButton;
 
 private slots:
     void Save();
     void Exit();
     void Load();
+    void ConfirmClearDatabase();
+    void ClearDatabase();
 };
 
 #endif // GALLERYCONFIG_H
diff --git a/mythtv/programs/mythfrontend/gallerydatabasehelper.cpp b/mythtv/programs/mythfrontend/gallerydatabasehelper.cpp
index 9288f8f..47a678d 100644
--- a/mythtv/programs/mythfrontend/gallerydatabasehelper.cpp
+++ b/mythtv/programs/mythfrontend/gallerydatabasehelper.cpp
@@ -653,3 +653,23 @@ QString GalleryDatabaseHelper::GetSortOrder()
 
     return sort;
 }
+
+
+
+/** \fn     GalleryDatabaseHelper::ClearDatabase()
+ *  \brief  Removes all contents from the gallery_directories and gallery_files tables.
+ *  \return void
+ */
+void GalleryDatabaseHelper::ClearDatabase()
+{
+    MSqlQuery query(MSqlQuery::InitCon());
+    query.prepare(QString("TRUNCATE gallery_directories;"));
+
+    if (!query.exec())
+        LOG(VB_GENERAL, LOG_ERR, MythDB::DBErrorMessage(query.lastError()));
+
+    query.prepare(QString("TRUNCATE gallery_files;"));
+
+    if (!query.exec())
+        LOG(VB_GENERAL, LOG_ERR, MythDB::DBErrorMessage(query.lastError()));
+}
diff --git a/mythtv/programs/mythfrontend/gallerydatabasehelper.h b/mythtv/programs/mythfrontend/gallerydatabasehelper.h
index 0038236..176cecb 100644
--- a/mythtv/programs/mythfrontend/gallerydatabasehelper.h
+++ b/mythtv/programs/mythfrontend/gallerydatabasehelper.h
@@ -34,6 +34,8 @@ public:
     void UpdateData(ImageMetadata *);
     void RemoveData(ImageMetadata *);
 
+    void ClearDatabase();
+
 private:
     void LoadDirectoryValues(MSqlQuery &, ImageMetadata *);
     void LoadFileValues(MSqlQuery &, ImageMetadata *);
-- 
1.7.10.2

